src/librustc_driver/driver.rs

```diff
@@ -826,6 +828,391 @@ pub fn phase_2_configure_and_expand<F>(sess: &Session,
+            write_traces_rec(file, &t.extent, total, depth + 1);
+            write!(file, "</div>\n").unwrap();
+        }
+    }
+
+    fn compute_counts_rec(counts: &mut HashMap<String,QueryMetric>, traces:&Vec<Rec>) {
+        for t in traces.iter() {
+            match t.effect {
+                Effect::QueryBegin(ref qmsg, ref _cc) => {
+                    let qcons = cons_of_query_msg(qmsg);
+                    let qm = match counts.get(&qcons) {
+                        Some(qm) =>
+                            QueryMetric{
+                                count:qm.count + 1,
+                                duration:qm.duration + t.duration
+                            },
+                        None => QueryMetric{
+                            count: 1,
+                            duration: t.duration
+                        }
+                    };
+                    counts.insert(qcons, qm);
+                }
+            }
+            compute_counts_rec(counts, &t.extent)
+        }
+    }
+
+    pub fn write_counts(count_file:&mut File, counts:&mut HashMap<String,QueryMetric>) {
+        use rustc::util::common::duration_to_secs_str;
+        use std::cmp::Ordering;
+
+        let mut data = vec![];
+        for (ref cons, ref qm) in counts.iter() {
+            data.push((cons.clone(), qm.count.clone(), qm.duration.clone()));
+        };
+        data.sort_by(|&(_,_,d1),&(_,_,d2)|
+                     if d1 > d2 { Ordering::Less } else { Ordering::Greater } );
+        for (cons, count, duration) in data {
+            write!(count_file, "{},{},{}\n",
+                   cons, count, duration_to_secs_str(duration)
+            ).unwrap();
+        }
+    }
+
+    pub fn write_traces(html_file:&mut File, counts_file:&mut File, traces:&Vec<Rec>) {
+        let mut counts : HashMap<String,QueryMetric> = HashMap::new();
+        compute_counts_rec(&mut counts, traces);
+        write_counts(counts_file, &mut counts);
+
+        let total : Duration = total_duration(traces);
+        write_traces_rec(html_file, traces, total, 0)
+    }
+
+    pub fn write_style(html_file:&mut File) {
+        write!(html_file,"{}", "
+body {
+    font-family: sans-serif;
+    background: black;
+}
+.trace {
+    color: black;
+    display: inline-block;
+    border-style: solid;
+    border-color: red;
+    border-width: 1px;
+    border-radius: 5px;
+    padding: 0px;
+    margin: 1px;
+    font-size: 0px;
+}
+.miss {
+    border-color: red;
+    border-width: 1px;
+}
+.extent-0 {
+    padding: 2px;
+}
+.important {
+    border-width: 3px;
+    font-size: 12px;
+    color: white;
+    border-color: #f77;
+}
+.hit {
+    padding: 0px;
+    border-color: blue;
+    border-width: 3px;
+}
+.eff {
+  color: #fff;
+  display: inline-block;
+}
+.frc {
+  color: #7f7;
+  display: inline-block;
+}
+.dur {
+  display: none
+}
+").unwrap();
+    }
+}
+
+
+pub fn profile_queries_thread(r:Receiver<ProfileQueriesMsg>) {
```

Comment:
I think I'd also prefer to move *this* code out into another file. I feel like a `profile/mod.rs` might be a nice place for this, with `trace` as a sub-module.